src/TypeSpec.Extension/Emitter.Csharp/src/lib/model.ts

```diff
@@ -667,40 +666,65 @@ export function getInputType(
             logger.info(
                 `No specified type for discriminator property '${model.DiscriminatorPropertyName}'. Assume it is a string.`
             );
-            const discriminatorProperty = {
+            const discriminatorProperty: InputModelProperty = {
                 Name: model.DiscriminatorPropertyName,
                 SerializedName: model.DiscriminatorPropertyName,
                 Description: "Discriminator",
                 IsRequired: true,
                 IsReadOnly: false,
-                IsNullable: false,
                 Type: {
                     Name: "string",
                     Kind: InputTypeKind.String,
                     IsNullable: false
                 } as InputPrimitiveType,
                 IsDiscriminator: true
-            } as InputModelProperty;
+            };
             // put default discriminator property as the first property to keep previous behavior
             outputProperties.unshift(discriminatorProperty);
         }
     }
-    function getInputModelBaseType(m?: Model): InputModelType | undefined {
-        if (!m) {
-            return undefined;
-        }
 
-        // Arrays and dictionaries can't be a base type
-        if (m.indexer) {
-            return undefined;
+    function getInputModelBaseType(
+        m: Model
+    ): [InputModelType | undefined, InputDictionaryType | undefined] {
+        const baseModel = m.baseModel;
+        const sourceModel = m.sourceModel;
+
+        // we cannot have both `extends` and `is`, therefore only one of baseModel and sourceModel can be defined
+        if (
+            sourceModel !== undefined &&
```

Comment:
```suggestion
            !sourceModel &&
```

Reply:
well when I wrote this, I kind of did not figure out how to simplify this expression and ended up writing this
I will update